src/inmanta/db/versions/v7.py

```diff
@@ -15,14 +15,72 @@
 
     Contact: code@inmanta.com
 """
-from asyncpg import Connection
+import datetime
+from functools import reduce
+from typing import Iterator, List, Optional
+
+from asyncpg import Connection, Record
 
 DISABLED = False
 
 
 async def update(connection: Connection) -> None:
     await connection.execute(
         """
-CREATE INDEX resourceaction_environment_version_started_index ON resourceaction(environment,version,started DESC);
+        CREATE INDEX resourceaction_environment_version_started_index ON resourceaction(environment,version,started DESC);
         """
     )
+    await enforce_unique_agent_instances(connection)
+
+
+async def enforce_unique_agent_instances(connection: Connection) -> None:
+    async with connection.transaction():
+        await merge_agent_instances(connection)
+        await connection.execute(
+            """
+            ALTER TABLE public.agentinstance ADD CONSTRAINT agentinstance_unique UNIQUE (tid, process, name);
+            """
+        )
+
+
+async def merge_agent_instances(connection: Connection) -> None:
```

Comment:
would it not be easier to something like this:

https://www.postgresqltutorial.com/how-to-delete-duplicate-rows-in-postgresql/

```
DELETE FROM
    public.agentinstance a
        USING public.agentinstance b
WHERE
    a.id < b.id
    AND a.tid = b.tid and a.process = b.process and a.name=b.name;
```

Iterating over an SQL result to generate many SQL statements is usually slow and often leads to unexpected problems in my limited experience.

Reply:
This looks interesting. I'll need to extend this to correctly set the expired state as well, unless it's all right to just set it to now (see my comment below)?

Reply:
Or even to just leave it as is.

Reply:
I would leave it as it is.

Reply:
afa19f9a